modules/client/src/main/java/org/apache/ignite/internal/client/table/AbstractClientView.java

```diff
@@ -135,20 +124,29 @@ public Cursor<T> query(@Nullable Transaction tx, @Nullable Criteria criteria, @N
     public CompletableFuture<AsyncCursor<T>> queryAsync(
             @Nullable Transaction tx,
             @Nullable Criteria criteria,
-            CriteriaQueryOptions opts
+            @Nullable CriteriaQueryOptions opts
     ) {
+        CriteriaQueryOptions opts0 = opts == null ? CriteriaQueryOptions.DEFAULT : opts;
+
         return tbl.getLatestSchema()
                 .thenCompose((schema) -> {
                     SqlSerializer ser = createSqlSerializer(tbl.name(), schema.columns(), criteria);
+
+                    Statement statement = new ClientStatementBuilder().query(ser.toString()).pageSize(opts0.pageSize()).build();
                     Session session = new ClientSessionBuilder(tbl.channel()).build();
 
-                    return session.executeAsync(tx, ser.toString(), ser.getArguments())
-                            .thenApply(resultSet -> {
+                    return session.executeAsync(tx, statement, ser.getArguments())
+                            .<AsyncCursor<T>>thenApply(resultSet -> {
                                 ResultSetMetadata meta = resultSet.metadata();
 
                                 assert meta != null : "Metadata can't be null.";
 
-                                return new QueryCriteriaAsyncCursor<>(resultSet, queryMapper(meta, schema), session::close);
+                                return new QueryCriteriaAsyncCursor<>(resultSet, queryMapper(meta, schema), session::closeAsync);
+                            })
+                            .exceptionally(th -> {
+                                session.closeAsync();
+
+                                throw new CompletionException(unwrapCause(th));
```

Comment:
Need to cover the case by test, check that session will be closed and check returned exception

Reply:
Added `ItCriteriaQueryTest#testSessionClosing`